eth-providers/src/utils/BlockCache.ts

```diff
@@ -7,44 +7,51 @@ export interface BlockToHashesMap {
 }
 
 export interface CacheInspect {
-  extraBlockCount: number;
+  maxCachedBlocks: number;
   cachedBlocksCount: number;
   cachedBlocks: string[];
   allBlockToHash: Record<string, string[]>;
   allHashToBlock: Record<string, number>;
 }
 
 export class BlockCache {
-  blockTxHashes: BlockToHashesMap;
-  allTxHashes: HashToBlockMap;
-  extraBlockCount: number;
-
-  constructor(extraBlockCount: number = 10) {
-    this.blockTxHashes = {};
-    this.allTxHashes = {};
-    this.extraBlockCount = extraBlockCount;
+  blockToHashes: BlockToHashesMap;
+  hashToBlocks: HashToBlockMap;
+  maxCachedBlocks: number;
+
+  constructor(maxCachedBlocks: number = 200) {
+    this.blockToHashes = {};
+    this.hashToBlocks = {};
+    this.maxCachedBlocks = maxCachedBlocks;
   }
 
+  // automatically preserve a sliding window of ${maxCachedBlocks} blocks
   addTxsAtBlock(blockNumber: number, txHashes: string[]): void {
-    txHashes.forEach((h) => (this.allTxHashes[h] = blockNumber));
-    this.blockTxHashes[blockNumber] = txHashes;
+    txHashes.forEach((h) => (this.hashToBlocks[h] = blockNumber));
+    this.blockToHashes[blockNumber] = txHashes;
+
+    const cachedBlocksCount = Object.keys(this.blockToHashes).length;
+    if (cachedBlocksCount > this.maxCachedBlocks) {
+      const blockToRemove = blockNumber - this.maxCachedBlocks;
+      this._removeBlock(blockToRemove);
+    }
   }
 
-  handleFinalizedBlock(blockNumber: number): void {
-    const blockToRemove = blockNumber - this.extraBlockCount;
-    this.blockTxHashes[blockToRemove]?.forEach((h) => delete this.allTxHashes[h]);
-    delete this.blockTxHashes[blockToRemove];
+  // if block exist in cache, remove it, otherwise do nothing
+  _removeBlock(blockToRemove: number): void {
+    this.blockToHashes[blockToRemove]?.forEach((h) => delete this.hashToBlocks[h]);
+    delete this.blockToHashes[blockToRemove];
```

Comment:
check `undefined` before deleting.

Reply:
check which one `undefined`?

Reply:
`this.blockToHashes` is always defined

Reply:
`this.blockToHashes`  may be not consecutive numbers, such as the `ws` connect interrupt

Reply:
deleting empty stuff is ok
```
const x = {}
delete x.aaaaa.  // => true
```

Reply:
please see my comment here https://github.com/AcalaNetwork/bodhi.js/pull/367/files#diff-505de5c3567ec34cbf3e2ba1a24e5356e13a445a3a53a18c152f89419af660e5R40

Reply:
also from here it should already made sure target is defined
https://github.com/AcalaNetwork/bodhi.js/pull/367/files#diff-505de5c3567ec34cbf3e2ba1a24e5356e13a445a3a53a18c152f89419af660e5R35-R36

Reply:
LGTM